File: js/five.js
// 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

// What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

function get_factors(num){
  var factors = [];
  var max_factor = Math.sqrt(num);
  for(var i = 1; i <= max_factor; i++){
    if(num % i == 0 && factors.indexOf(i) < 0){
      factors.push(i);
      if(num / i != i){
        factors.push(num / i);
      }
    }
  }
  return factors.sort(function(a,b){return a - b});
}

function get_prime_factors(num){
  var primes = [];
  for(var i = 1; i < num; i++){
    if(get_factors(i).length == 2){
      primes.push(i);
    }
  }
  return primes;
}

function get_minimum_product(num){
  var prime_factors = get_prime_factors(num);
  var product = 1;
  for(var i = 0; i < prime_factors.length; i++){
    var prime = prime_factors[i];
    var x = 1;
    while(Math.pow(prime, x + 1) < num){
      x++;
    }
    var product = product * Math.pow(prime, x);
  }
  return product;
}

File: rb/five.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
class Integer
  def get_factors 
    factors = []
    (1..Math.sqrt(self)).each {|num|
      if(self % num == 0 && !factors.include?(num))
        factors.push(num)
        factors.push(self/num)
      end
    }
    factors
  end
  def get_primes_factors
    primes = []
    (1..self).each {|num|
      if num.get_factors.length == 2
        primes.push(num)
      end
    }
    primes
  end
end


def smallest_divisible num
  primes = num.get_primes_factors
  products = []
  primes.each {|prime|
    next if prime == 1
    i = 1
    while prime**i < num
      i += 1
    end
    i -= 1
    products.push(prime**i)
  }
  products.inject {|a,b| a * b}
end

puts smallest_divisible 20